src/java.desktop/macosx/native/libosxapp/JNIUtilities.h

```diff
@@ -0,0 +1,161 @@
+     } \
+     LOG_NULL(dst_var, name); \
+     CHECK_NULL(dst_var);
+
+#define DECLARE_METHOD(dst_var, cls, name, signature) \
+     static jmethodID dst_var = NULL; \
+     GET_METHOD(dst_var, cls, name, signature);
+
+#define GET_METHOD_RETURN(dst_var, cls, name, signature, ret) \
+     if (dst_var == NULL) { \
+         dst_var = (*env)->GetMethodID(env, cls, name, signature); \
+     } \
+     LOG_NULL(dst_var, name); \
+     CHECK_NULL_RETURN(dst_var, ret);
+
+#define DECLARE_METHOD_RETURN(dst_var, cls, name, signature, ret) \
+     static jmethodID dst_var = NULL; \
+     GET_METHOD_RETURN(dst_var, cls, name, signature, ret);
+
+#define GET_STATIC_METHOD(dst_var, cls, name, signature) \
+     if (dst_var == NULL) { \
+         dst_var = (*env)->GetStaticMethodID(env, cls, name, signature); \
+     } \
+     LOG_NULL(dst_var, name); \
+     CHECK_NULL(dst_var);
+
+#define DECLARE_STATIC_METHOD(dst_var, cls, name, signature) \
+     static jmethodID dst_var = NULL; \
+     GET_STATIC_METHOD(dst_var, cls, name, signature);
+
+#define GET_STATIC_METHOD_RETURN(dst_var, cls, name, signature, ret) \
+     if (dst_var == NULL) { \
+         dst_var = (*env)->GetStaticMethodID(env, cls, name, signature); \
+     } \
+     LOG_NULL(dst_var, name); \
+     CHECK_NULL_RETURN(dst_var, ret);
+
+#define DECLARE_STATIC_METHOD_RETURN(dst_var, cls, name, signature, ret) \
+     static jmethodID dst_var = NULL; \
+     GET_STATIC_METHOD_RETURN(dst_var, cls, name, signature, ret);
+
+/********        GET FIELD SUPPORT    *********/
+
+
+#define GET_FIELD(dst_var, cls, name, signature) \
+     if (dst_var == NULL) { \
+         dst_var = (*env)->GetFieldID(env, cls, name, signature); \
+     } \
+     LOG_NULL(dst_var, name); \
+     CHECK_NULL(dst_var);
+
+#define DECLARE_FIELD(dst_var, cls, name, signature) \
+     static jfieldID dst_var = NULL; \
+     GET_FIELD(dst_var, cls, name, signature);
+
+#define GET_FIELD_RETURN(dst_var, cls, name, signature, ret) \
+     if (dst_var == NULL) { \
+         dst_var = (*env)->GetFieldID(env, cls, name, signature); \
+     } \
+     LOG_NULL(dst_var, name); \
+     CHECK_NULL_RETURN(dst_var, ret);
+
+#define DECLARE_FIELD_RETURN(dst_var, cls, name, signature, ret) \
+     static jfieldID dst_var = NULL; \
+     GET_FIELD_RETURN(dst_var, cls, name, signature, ret);
+
+#define GET_STATIC_FIELD_RETURN(dst_var, cls, name, signature, ret) \
+     if (dst_var == NULL) { \
+         dst_var = (*env)->GetStaticFieldID(env, cls, name, signature); \
+     } \
+     LOG_NULL(dst_var, name); \
+     CHECK_NULL_RETURN(dst_var, ret);
+
+#define DECLARE_STATIC_FIELD_RETURN(dst_var, cls, name, signature, ret) \
+     static jfieldID dst_var = NULL; \
+     GET_STATIC_FIELD_RETURN(dst_var, cls, name, signature, ret);
+
+/*********       EXCEPTION_HANDLING    *********/
+
+#define CHECK_EXCEPTION() \
+    if ((*env)->ExceptionOccurred(env) != NULL) { \
+        (*env)->ExceptionClear(env); \
```

Comment:
Why an exception is cleared here? Is it really match the behavior before the fix? I guess before the fix various JNF methods thrown the objective-c exceptions which are usually handled by the JNF_COCOA_ENTER/EXIT or in the NSApplicationAWT.runAWTLoopWithApp()

Reply:
The behaviour may not be identical but it is non-identical in theoretical fatal cases.
We are actually better off here. I am seeing fewer JNI warning as a result.
Also NS exceptions per Apple should result in termination of the process.
I don't think JNF ever did the right thing.
So as I am sure you know this is only ever going to come into play if there is some JNI
error which indicates an internal error. I don't think we are any worse off here.

Reply:
I see a mixing approach here, on other platforms like on windows we:
 - Use such macro in the initid() method in the static initializers and any errors are fatal in that case. And we usually use the macro instead of functions, because it is possible to check and return from the initid().
 - If some exception like OOM has occurred on the toolkit thread we wrap it to the std::bad_alloc end stop the application.

The JNF solution on macOS merges two solutions above, so we call JNI methods in any place and the library transforms all possible java exceptions to the NSExceptions, and back. And it was quite useful, much better than the list of macro.

Reply:
I did analyse what seems to be happening with the existing JNF code and am comfortable with it.
The use of NSExceptions in that case is really against the grain of what Apple say for NSExceptions that
they are meant for terminal conditions, so it is a bit weird to just use them to "carry" a Java Exception.
However if we do see one on return (the EXIT code yet to be updated) then we would still turn it into a Java exception.
I am against converting a Java Exception into an NSException then back into a Java exception which seems
unnecessary.